core/rfb.js

```diff
@@ -1496,6 +1499,76 @@ export default class RFB extends EventTargetMixin {
         return true;
     }
 
+    _stringToAsciiByteArray(str) {
+        let bytes = new Uint8Array(str.length);
+        for (let i = 0; i < str.length; ++i) {
+            bytes[i] = str.charCodeAt(i);
+        }
+        return bytes;
+    }
+
+    _negotiateARDAuth() {
+
+        if (this._rfbCredentials.username === undefined ||
+            this._rfbCredentials.password === undefined) {
+            this.dispatchEvent(new CustomEvent(
+                "credentialsrequired",
+                { detail: { types: ["username", "password"] } }));
+            return false;
+        }
+
+        if (this._sock.rQwait("read ard", 4)) { return false; }
+
+        let generator = this._sock.rQshiftBytes(2);   // DH base generator value
+
+        let keyLength = this._sock.rQshift16();
+
+        if (this._sock.rQwait("read ard keylength", keyLength*2, 4)) { return false; }
+
+        let primeBytes = this._sock.rQshiftBytes(keyLength);  // predetermined prime modulus
+
+        let serverPubKeyBytes = this._sock.rQshiftBytes(keyLength); // other party's public key
+
+        let prime = new forge.jsbn.BigInteger(forge.util.createBuffer(primeBytes).toHex(), 16);
+
+        let clientPrivKey = new forge.jsbn.BigInteger(forge.util.createBuffer(forge.random.generate(keyLength)).toHex(), 16);
+
+        let ghd = new forge.jsbn.BigInteger(generator);
+        let clientPubKey = ghd.modPow(clientPrivKey, prime);
+
+        let serverPubKey = new forge.jsbn.BigInteger(forge.util.createBuffer(serverPubKeyBytes).toHex(), 16);
+
+        let sharedKey = serverPubKey.modPow(clientPrivKey, prime);
+
+        let md5Shared = forge.md.md5.create().update(forge.util.hexToBytes(sharedKey.toString(16))).digest();
+
+        let userPassword = this._rfbCredentials.username + '\0'
+                + forge.util.createBuffer(forge.random.generate(32)).toHex().substring(0, 63-this._rfbCredentials.username.length) +
+                this._rfbCredentials.password + '\0'
+                + forge.util.createBuffer(forge.random.generate(32)).toHex().substring(0, 63-this._rfbCredentials.password.length);
+
+        let packedUserPassword = forge.util.encodeUtf8(userPassword);
+
+        let cipher = forge.cipher.createCipher('AES-ECB', md5Shared);
+
+        cipher.mode.pad = false;
+        cipher.start();
+        cipher.update(forge.util.createBuffer(packedUserPassword));
+        cipher.finish();
+
+        let data = this._stringToAsciiByteArray(cipher.output.data);
+        this._sock.send(data);
+
+        let pubBytes = clientPubKey.toByteArray();
+        if (pubBytes.length==129) {
```

Comment:
This seems odd to hard code. Doesn't it depend on the key length?

Reply:
> This seems odd to hard code. Doesn't it depend on the key length?

For some reason sometimes the client pub key comes out at 129 bytes with an extra 0 at the beginning, i'm not sure why it's happening but this fixes that case and makes it work all the time.

Reply:
Right, that is common. However specifically `129` seems wrong. `clientPubKey` is `keyLength` bytes large, not always 128. I'm assuming in your test `keyLength` happens to be 128 bytes?